compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -10690,6 +10730,25 @@ private STNode parseRetryBody() {
         }
     }
 
+    private STNode parseOnFailClause() {
+        STNode onKeyword = parseOnKeyword();
+        STNode failKeyword;
+        STToken nextToken = peek();
+        if (nextToken.kind == SyntaxKind.FAIL_KEYWORD) {
+            failKeyword = parseFailKeyword();
+        } else {
+            // No other cases for now
+            failKeyword = STNodeFactory.createEmptyNode();
+        }
```

Comment:
According to the grammar, fail keyword is not optional.

Reply:
Fixed with 3189bc6c6a9ee435ca57fa863f10d8e4f357ab6b